### PR TITLE
[Package Signing] Add Certificates Validator

### PR DESCRIPTION
**Pull request by loic-sharma**
This Validator integrates with the validation Orchestrator to kick off one or more `ValidateCertificate` jobs for signed packages. The `PackageCertificatesValidator` will wait until the `ValidateCertificate` jobs completes before finalizing the overall validation.

Fixes https://github.com/NuGet/Engineering/issues/786.

### CODE REVIEW

File: tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidator.cs [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Moq;
+using NuGet.Services.Validation.Orchestrator;
+using NuGet.Services.Validation.PackageCertificates;
+using Xunit;
+
+namespace NuGet.Services.Validation.PackageSigning
+{
+    public class PackageCertificatesValidatorFacts
```
Comment (joelverhagen) [inline tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidator.cs]: >PackageCertificatesValidatorFacts [](start = 17, length = 33)

Rename this file to match the class name.

File: src/NuGet.Services.Validation.Orchestrator/PackageSigning/PackageSigningValidator.cs [hunk 2]
```diff
@@ -61,7 +61,7 @@ public class PackageSigningValidator : IValidator
                 // message has been duplicated.
                 _logger.LogError(
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/PackageSigning/PackageSigningValidator.cs]: >LogError [](start = 24, length = 8)

Is this really an error? This feels more like a warning.

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+            var signatures = await FindSignaturesAsync(request);
+            var invalidSignatures = signatures.Where(SignatureShouldBeInvalidated).ToList();
+
+            if (invalidSignatures.Any())
+            {
+                var package = await FindPackageSigningStateAsync(request);
+
+                InvalidatePackageSignatures(request, package, invalidSignatures);
+
+                return await TryUpdateValidationStatus(request, status, ValidationStatus.Failed);
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: >TryUpdateValidationStatus [](start = 29, length = 25)

`GetStatusAsync` is intended to be read-only (no concerns of concurrency, could be run on the secondary, etc etc). Can we make this read-only?

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+                return await TryAddValidatorStatus(request, status, ValidationStatus.Failed);
+            }
+
+            // Find the signatures used to sign the package and see if any certificates known to be revoked
+            // invalidate any of these signatures. Note that the certificate statuses' may be stale at this point,
+            // however, revoked certificates remain revoked forever. All other certificates will be checked later.
+            var signatures = await FindSignaturesAsync(request);
+            var invalidSignatures = signatures.Where(SignatureShouldBeInvalidated).ToList();
+
+            if (invalidSignatures.Any())
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: >invalidSignatures [](start = 16, length = 17)

This looks similar to code above in `GetStatusAsync` -- can we share the logic in a function?

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+        /// Check whetehr all certificate validations for the given validation request are finished.
+        /// </summary>
+        /// <param name="request">The validation request that started the certificate validations.</param>
+        /// <returns>Whether the certificate validations are ALL finished.</returns>
+        private Task<bool> AllCertificateValidationsAreFinishedAsync(IValidationRequest request)
+        {
+            return _validationContext
+                    .CertificateValidations
+                    .Where(v => v.ValidationId == request.ValidationId)
+                    .AllAsync(v => v.IsFinished());
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: I don't think `IsFinished` can be evaluated in SQL Server right? This either needs to be an expression or after the query is materialized.

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs (hunk 3 above)
Comment (joelverhagen) [reply src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: Oh and most importantly I was hoping this should be callable from admin panel without concerns of data changes.

---
In reply to: [143584031](https://github.com/NuGet/NuGet.Jobs/pull/225#discussion_r143584031) [](ancestors = 143584031)

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+                // with a status of "InGracePeriod" whose "SignedAt" time is less than all of the signature's
+                // certificates' "StatusUpdateTime". The resulting signatures' statuses should be promoted
+                // to "Valid". Currently, certificate grace periods are not handled, thus all signatures
+                // should already be in the "Valid" state.
+                return await TryAddValidatorStatus(request, status, ValidationStatus.Succeeded);
+            }
+        }
+
+        /// <summary>
+        /// Check whetehr all certificate validations for the given validation request are finished.
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: >whetehr [](start = 18, length = 7)

whether

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+                var timeAgo = certificate.LastVerificationTime.Value - DateTime.UtcNow;
+
+                return timeAgo >= _certificateValidationCacheSpan;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Enqueues certificate verifications and add <see cref="CertificateValidation"/> entities
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: >Enqueues [](start = 12, length = 8)

To have consistent voice of your other comments: "Enqueue" instead of "Enqueues"

File: src/NuGet.Services.Validation.Orchestrator/PackageSigning/PackageSigningValidator.cs (hunk 2 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/PackageSigning/PackageSigningValidator.cs]: Sure. I'll move all other `LogError` to `LogWarning` for database concurrency issues.

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+
+namespace NuGet.Services.Validation.PackageCertificates
+{
+    public class PackageCertificatesValidator : IValidator
+    {
+        private IValidationEntitiesContext _validationContext;
+        private IValidatorStateService _validatorStateService;
+        private ICertificateVerifier _certificateVerifier;
+        private TimeSpan _certificateValidationCacheSpan;
+        private ILogger<PackageCertificatesValidator> _logger;
```
Comment (xavierdecoster) [inline src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: All of the above fields can be made `readonly`.

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs (hunk 5 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: Good catch. This wasn't caught by unit tests, but this did reproduce when I made a test console app for integration testing.

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs (hunk 4 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: Removed this logic from `GetStatusAsync`

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs (hunk 3 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: Discussed this with Joel offline at great lengths. Ideally, the `GetStatusAsync` should be *completely* read-only. However, achieving this would complicate the pipeline a little. I personally do not think that the trade-off is worth it as of right now. If this becomes an issue, we can move the core of `GetStatusAsync` to its own job, and have that update the `ValidatorStatus` when all of the `CertificateValidation`s complete.

**Comment by loic-sharma** [comment]
Addressed all feedback, please review :)

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+
+            return certificates.ToList();
+        }
+
+        /// <summary>
+        /// Decide whether or not to revalidate the given certificate.
+        /// </summary>
+        /// <param name="certificate">The certificate that may be revalidated.</param>
+        /// <returns>Whether the certificate should be revalidated.</returns>
+        private bool ShouldValidateCertificate(Certificate certificate)
```
Comment (agr) [inline src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: `static`

File: tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs [hunk 10]
```diff
@@ -0,0 +1,1098 @@
...
+                // The validation should fail if the package was signed after the certificate invaliditiy begins.
+                yield return new object[]
+                {
+                    ValidationStatus.Failed, DateTime.UtcNow.AddDays(-1), DateTime.UtcNow.AddDays(-10)
+                };
+            }
+
+            [Theory]
+            [MemberData(nameof(RevokedCertificatesFailsValiationOnlyIfRevokedBeforeSignatureTimeData))]
+            public async Task RevokedCertificatesFailsValiationOnlyIfRevokedBeforeSignatureTime(
```
Comment (scottbommarito) [inline tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs]: `Valiation` -> `Validation`

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/CertificateVerifier.cs [hunk 11]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using NuGet.Services.ServiceBus;
+
+namespace NuGet.Services.Validation.PackageCertificates
+{
+    public class CertificateVerifier : ICertificateVerifier
```
[the remark this answers is not in the document]
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/PackageCertificates/CertificateVerifier.cs]: Name change sounds good. We have a generic service bus serialization class, but not an enqueuer class.

File: tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs (hunk 10 above)
Comment (loic-sharma) [reply tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs]: Nice catch :)

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+            ICertificateVerifier certificateVerifier,
+            ILogger<PackageCertificatesValidator> logger,
+            TimeSpan? certificateStatusCacheTime = null)
+        {
+            _validationContext = validationContext ?? throw new ArgumentNullException(nameof(validationContext));
+            _validatorStateService = validatorStateService ?? throw new ArgumentNullException(nameof(validatorStateService));
+            _certificateVerifier = certificateVerifier ?? throw new ArgumentNullException(nameof(certificateVerifier));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            _certificateValidationCacheSpan = certificateStatusCacheTime ?? TimeSpan.FromDays(1);
```
[the remark this answers is not in the document]
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: I suppose `TimeSpan.Zero` could be used if one intended to never cache certificates' statuses. But yes, a negative TimeSpan would be nonsensical :)

File: tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs [hunk 13]
```diff
@@ -0,0 +1,936 @@
...
+namespace NuGet.Services.Validation.PackageSigning
+{
+    public class PackageCertificatesValidatorFacts
+    {
+        private const int PackageKey = 1001;
+        private const string PackageId = "NuGet.Versioning";
+        private const string PackageNormalizedVersion = "4.3.0.0-ALPHA+git";
+        private static readonly Guid ValidationId = new Guid("fb9c0bac-3d4d-4cc7-ac2d-b3940e15b94d");
+        private static readonly Guid OtherValidationId = new Guid("6593BD33-ABC0-4049-BDCF-915807F1D2B3");
+        private const string NupkgUrl = "https://example/nuget.versioning/4.3.0/package.nupkg";
```
Comment (dtivel) [inline tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs]: Consider using [`unit.test`](https://en.wikipedia.org/wiki/.test) instead of `example`.

File: tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs [hunk 14]
```diff
@@ -0,0 +1,936 @@
...
+                    new CertificateStatus?[] { CertificateStatus.Unknown },
+                };
+
+                yield return new object[]
+                {
+                    ValidationStatus.Succeeded,
+                    new CertificateStatus?[] { CertificateStatus.Good, CertificateStatus.Unknown, CertificateStatus.Good },
+                };
+
+                // The validator should be resilient to no certificates. This should be an impossible state.
```
Comment (dtivel) [inline tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs]: Unsigned packages will have `ValidationStatus.Succeeded` and no certificates.  See [here](https://github.com/NuGet/NuGet.Jobs/pull/225/files#diff-6d7330dc27fcee16ad09316d52d642e3R112).  How is it impossible?

File: src/NuGet.Services.Validation.Orchestrator/NuGet.Services.Validation.Orchestrator.csproj [hunk 15]
```diff
@@ -48,6 +48,9 @@
     <Compile Include="Error.cs" />
     <Compile Include="IValidatorStateService.cs" />
     <Compile Include="Job.cs" />
+    <Compile Include="PackageCertificates\CertificateVerifier.cs" />
+    <Compile Include="PackageCertificates\ICertificateVerifier.cs" />
```
Comment (dtivel) [inline src/NuGet.Services.Validation.Orchestrator/NuGet.Services.Validation.Orchestrator.csproj]: I think this project won't compile.  These files have been renamed, but the .csproj doesn't contain the new file names.

File: tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs [hunk 16]
```diff
@@ -0,0 +1,936 @@
...
+                // signing if the package was signed AFTER the revocation date.
+                yield return new object[]
+                {
+                    ValidationStatus.Incomplete,
+                    new CertificateStatus?[] { null },
+                };
+
+                yield return new object[]
+                {
+                    ValidationStatus.Incomplete,
```
Comment (dtivel) [inline tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs]: I'm not sure if it's just an issue with this validation status or what, but "incomplete" implies non-finality.  However, if a certificate has been revoked, that's a final state for verification; there's nothing unfinished about the validation.

It seems like the validation status should be failed.

File: tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs [hunk 17]
```diff
@@ -0,0 +1,936 @@
...
+
+                _certificateVerifier.Verify(v => v.StartVerificationAsync(It.IsAny<IValidationRequest>(), It.IsAny<Certificate>()), Times.Never);
+                _validationContext.Verify(c => c.SaveChangesAsync(), Times.Once);
+
+                Assert.Equal(ValidationStatus.Failed, actual);
+                Assert.Equal(ValidationStatus.Failed, validatorStatus.State);
+            }
+
+            [Fact]
+            public async Task OnRevalidationAllNonRevokedCertificatesAreVerified()
```
Comment (dtivel) [inline tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs]: I may have missed it, but do you have a test verifying that if a package has multiple signatures and one signature is with a revoked certificate, that revalidating non-revoked certificates won't mark the package as valid?

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+            _validationContext = validationContext ?? throw new ArgumentNullException(nameof(validationContext));
+            _validatorStateService = validatorStateService ?? throw new ArgumentNullException(nameof(validatorStateService));
+            _certificateVerifier = certificateVerifier ?? throw new ArgumentNullException(nameof(certificateVerifier));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            _certificateValidationCacheSpan = certificateStatusCacheTime ?? DefaultCertificateStatusCacheTime;
+
+            if (_certificateValidationCacheSpan < TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException("The certificate status cache time must be a positive value", nameof(certificateStatusCacheTime));
```
[the remark this answers is not in the document]
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: Weird - I could have sworn the parameter name came after the message in exceptions' constructors. I'll fix this :)

File: src/NuGet.Services.Validation.Orchestrator/NuGet.Services.Validation.Orchestrator.csproj (hunk 15 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/NuGet.Services.Validation.Orchestrator.csproj]: My bad. I've been doing a bunch of csproj hacks (replacing PackageReferences with ProjectReferences) to work around the current code signing issues, and I've been trying to avoid committing those hacks in.

File: tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs [hunk 19]
```diff
@@ -0,0 +1,936 @@
...
+                        RevocationTime = DateTime.MaxValue,
+                    };
+
+                    var certificateValidation = new CertificateValidation
+                    {
+                        ValidationId = ValidationId,
+                        Status = certificateStatus,
+                        Certificate = certificate
+                    };
+
```
Comment (loic-sharma) [inline tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs]: Nit: extra space

File: tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs (hunk 14 above)
Comment (loic-sharma) [reply tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs]: This comment is confusing. This test ensures that **signed** packages are properly labelled by `GetStatusAsync` as succeeded/failed/incomplete based off the status of their certificate validations. A **signed** package should have at least 1 certificate, hence why this state should be "impossible". Note that this validator runs on packages whose signatures have passed the client's signature validations, as per the `ExtractAndValidatePackageSignature` job. Thus, it is reasonable for this validator to assume that a signature will have at least one certificate.

TLDR: The comment is misleading, I will improve it :)

File: tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs (hunk 17 above)
Comment (loic-sharma) [reply tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs]: Interesting. I do not have any revalidation tests, I will see what additional corner cases I can come up with!

Side-note: once the `ValidateCertificate` job is running, I will add thorough integration testing between the `CertificatesValidator` and the `ValidateCertificate` job. This will be done in a separate pull request.

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: I found the source of my confusion. [`ArgumentException`](https://msdn.microsoft.com/en-us/library/sxykka64(v=vs.110).aspx) has the following signature:

```
public ArgumentException(string message, string paramName);
```

That's a pretty serious mistake by the .NET team 😢

File: tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs (hunk 16 above)
Comment (loic-sharma) [reply tests/NuGet.Services.Validation.Orchestrator.Tests/PackageCertificates/PackageCertificatesValidatorFacts.cs]: A validation will eventually fail if there are one or more revoked certificates - HOWEVER, the validator is designed to wait until ALL certificate validations have completed. If a certificate validation hasn't finished (aka a `null` `CertificateStatus`), the validator will say that the validation is `Incomplete`. The orchestrator will continue to ask the `CertificateValidator` for the latest status on the validations - and the `CertificateValidator` will eventually return `Failed` instead of `Incomplete`.

Why should the validator wait? A package may have multiple more than one invalid certificate. If the validator doesn't wait until all certificates are validated, it may miss invalid certificates and return an incorrect view on the package.

I will add a comment explaining this behavior :)

**Comment by loic-sharma** [comment]
Updated to use latest version of the validation context, rebased to the new feature branch `dev-package-signing`, and addressed all feedback.

*NOTE*: The spec has changed! This pull request does not reflect the latest changes. I will implement the changes tomorrow.

File: src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs [hunk 20]
```diff
@@ -0,0 +1,380 @@
...
+        /// </summary>
+        /// <param name="validationContext">The persisted validation context.</param>
+        /// <param name="validatorStateService">The service used to persist this validator's state.</param>
+        /// <param name="certificateVerifier">The verifier used to verify individual certificates asynchronously.</param>
+        /// <param name="certificateStatusCacheTime">How long certificates' statuses should be cached before revalidating. Defaults to 1 day.</param>
+        public PackageCertificatesValidator(
+            IValidationEntitiesContext validationContext,
+            IValidatorStateService validatorStateService,
+            ICertificateVerifier certificateVerifier,
+            ILogger<PackageCertificatesValidator> logger,
```
[the remark this answers is not in the document]
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/PackageCertificates/PackageCertificatesValidator.cs]: Fixed

**Comment by loic-sharma** [comment]
:bell: The validator is now feature complete. Please do one last round of review! :)
